### PR TITLE
Issue/655: Standard deviation and variance

### PR DESCRIPTION
**Pull request by allmywatts**
This PR implements `std` and `var` methods (see: #655) using the same method as used for `var_axis`, `std_axis`.
The variance is computed by flattening the array of N dimensions into a 1D array of the length `.len()` of the original array.

The return type is scalar A instead of Array<A, D:Smaller>

An attempt was made to refactor `var_axis` to use `var`, but this resulted in a performance regression so original implementation is kept and we accept some code duplication. :stuck_out_tongue: 

```rust
pub fn var_axis(&self, axis: Axis, ddof: A) -> Array<A, D::Smaller>
where
    A: Float + FromPrimitive,
    D: RemoveAxis,
{
    let mut sum_sq = Array::<A, D::Smaller>::zeros(self.dim.remove_axis(axis));

    for (lane, sum) in self.lanes(axis).into_iter().zip(sum_sq.iter_mut()) {
        *sum = lane.var(ddof);
    }
    sum_sq
}
```

Thanks @LukeMathWalker for the mentoring. :grinning:

### CODE REVIEW

**Comment by LukeMathWalker** [comment]
The main issue we had here was that `Zip` does not allow to iterator over two producers that yield items with different dimensionality (`0` for the output array and `1` for the lanes that we want to reduce).

**Comment by bluss** [comment]
I don't quite see the issue with using Zip here

**Comment by LukeMathWalker** [comment]
Sorry, it took me a while to get back to this - @bluss is indeed right, I was able to put it together using `Zip`. If you still have the attempts we made together @allmywatts at RustFest I'd be curious to understand what I was doing wrong there :thinking: 

I opened a PR against your branch with the `Zip` version @allmywatts - https://github.com/allmywatts/ndarray/pull/1
We might want to have a `var_unchecked`, as we did at the beginning, that we use from both `var` and `var_axis` to avoid redundant input checks in `var_axis`.

**Comment by hameerabbasi** [comment]
@allmywatts If you don't mind can I take over this PR and open a new one?

**Comment by allmywatts** [comment]
@hameerabbasi hey, sorry for the late reply, yes please go ahead.

**Comment by bluss** [comment]
This looks ready to merge, right?

File: src/numeric/impl_numeric.rs [hunk 1]
```diff
@@ -306,6 +294,67 @@ where
...
+        let n = A::from_usize(self.len()).expect("Converting length to `A` must not fail.");
+        assert!(
+            !(ddof < zero || ddof > n),
+            "`ddof` must not be less than zero or greater than the length of \
+             the axis",
+        );
+        let dof = n - ddof;
+        let mut mean = A::from_usize(0).expect("Converting 0 to `A` must not fail.");
+        let mut sum_sq = A::from_usize(0).expect("Converting 0 to `A` must not fail.");
+        for (count, x) in self.iter().enumerate() {
```
Comment (bluss) [inline src/numeric/impl_numeric.rs:327]: Just to note, iter is not an efficient way to iterate a "flattened" array, so this should potentially use unary Zip instead, for example, or `ArrayBase::fold`.

**Comment by bluss** [comment]
Super minor thing, but as long as there is a division and a checked type conversion inside the actual loop, I don't think `var_unchecked` with the purpose of saving input checks saves that much.

File: src/numeric/impl_numeric.rs [hunk 2]
```diff
@@ -306,6 +294,67 @@ where
...
+    /// # Example
+    ///
+    /// ```
+    /// use ndarray::{arr2, Axis};
+    ///
+    /// let a = arr2(&[[1., 2.],
+    ///                [3., 4.],
+    ///                [5., 6.]]);
+    ///
+    /// let a_flat = a.view().into_shape(6).expect("This must not fail.");
```
Comment (bluss) [inline src/numeric/impl_numeric.rs:348]: "This must not fail" does not add information, just use `.unwrap()`. I'm just afraid that the user will conclude some kinds of `into_shape()` (to 1D) always succeed, and they don't.

**Comment by bluss** [comment]
Superseded by #790

**Pull request closed without merging**